applications/OptimizationApplication/python_scripts/utilities/buffered_dict.py

```diff
@@ -0,0 +1,406 @@
+        buffer_data = self.__buffered_data[self.__GetBufferIndex(step_index)]
+        if key in buffer_data.keys():
+            raise RuntimeError(f"Trying to add a buffer value with key = \"{key}\" when already value exists for the key [ Existing value = {buffer_data[key]}]. BufferedDict:\n{self}")
+
+        # now check whether a sub item exists
+        if key in self.__sub_items.keys():
+            raise RuntimeError(f"Trying to add a buffer value with key = \"{key}\" when a subitem exists with the same key. BufferedDict:\n{self}")
+
+        # now add the buffer value
+        buffer_data[key] = value
+
+    def __AddSubItem(self, key: str, value: BufferedDict) -> None:
+        # first check if any of the buffered data has the same key.
+        # this is because, sub_items are valid for all step_indices.
+        if any([key in buffer_data.keys() for buffer_data in self.__buffered_data]):
+            raise RuntimeError(f"Trying to add a new a sub item with key = \"{key}\" when a value with the same key exists in buffered data. BufferedDict:\n{self}")
+
+        # now check if the item exists in the sub_items
+        if key in self.__sub_items.keys():
+            raise RuntimeError(f"Trying to add a new sub_item with key = \"{key}\" when already a sub item exists. BufferedDict:\n{self}")
+
+        # now add the sub_item
+        value.__parent = self
+        self.__sub_items[key] = value
+
+    def __SetBufferSize(self, buffer_size: int) -> None:
+        self.__buffer_index = 0
+        self.__buffered_data.clear()
+        for _ in range(buffer_size):
+            self.__buffered_data.append({})
+
+    def __GetBufferIndex(self, step_index: int) -> int:
+        if step_index >= self.GetBufferSize() or step_index < 0:
+            raise RuntimeError(f"Invalid step index. Allowed step indices are in [0, {self.GetBufferSize()}) [ StepIndex = {step_index} ]. BufferedDict:\n{self}")
+
+        return (self.__buffer_index - step_index) % self.GetBufferSize()
+
+    def __Info(self, step_index: int, tabbing: str) -> str:
+        info = "{"
+
+        if step_index == -1:
+            for i in range(self.GetBufferSize()):
+                info += f"\n{tabbing}\t--- Step = {i:2d} ---"
+                for key, value in self.__buffered_data[self.__GetBufferIndex(i)].items():
+                    info += f"\n{tabbing}\t\"{key}\": {value}"
+        else:
+            for key, value in self.__buffered_data[self.__GetBufferIndex(step_index)].items():
+                info += f"\n{tabbing}\t\"{key}\": {value}"
+
+        if step_index == -1:
+            info += f"\n{tabbing}\t--- Sub items ---"
+
+        next_tabbing = f"{tabbing}\t"
+        for sub_key, sub_value in self.__sub_items.items():
+            info += f"\n{tabbing}\t\"{sub_key}\": {sub_value.__Info(step_index, next_tabbing)}"
+
+        info += f"\n{tabbing}" + "}"
+        return info
+
+    def __getitem__(self, key: str) -> Any:
```

Comment:
again here we can simplify. Since you have implemented magic methods or dunder methods. there is absolutely no need for ClearStep function, one can do like buffered_data["step", 1].clear(). 
Using AdvanceStep and ClearStep functions is making complications without benefits.